[c10d] Monitored barrier: option to collect all failed ranks

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #55265 [c10d] Log API usage of monitored barrier
* #55197 [c10d] monitored_barrier: ensure all ranks pass or none do
* **#55010 [c10d] Monitored barrier: option to collect all failed ranks**
* #55009 [c10d] Enforce order of waited ranks in monitored barrier.

Follow up change to add a flag to provide an option for monitored barrier to collect all the failed ranks and then throw instead of just throwing on the first one. This is useful as now monitored barrier will be able to pick up on all hanging ranks instead of just one.

This is done by passing in a flag `wait_all_ranks=True`.

Differential Revision: [D27447787](https://our.internmc.facebook.com/intern/diff/D27447787/)

rohan-varma committed:
[c10d] Monitored barrier: option to collect all failed ranks

Follow up change to add a flag to provide an option for monitored barrier to collect all the failed ranks and then throw instead of just throwing on the first one. This is useful as now monitored barrier will be able to pick up on all hanging ranks instead of just one.

This is done by passing in a flag `wait_all_ranks=True`.

Differential Revision: [D27447787](https://our.internmc.facebook.com/intern/diff/D27447787/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[c10d] Monitored barrier: option to collect all failed ranks"


Follow up change to add a flag to provide an option for monitored barrier to collect all the failed ranks and then throw instead of just throwing on the first one. This is useful as now monitored barrier will be able to pick up on all hanging ranks instead of just one.

This is done by passing in a flag `wait_all_ranks=True`.

Differential Revision: [D27447787](https://our.internmc.facebook.com/intern/diff/D27447787/)

[ghstack-poisoned]

SciPioneer commented on torch/lib/c10d/ProcessGroupGloo.cpp:
```diff
@@ -115,7 +115,12 @@ using steady_clock_time_point =
+    bool waitAllRanks) {
+  if (waitAllRanks) {
+    // See Note in monitoredBarrier
+    return timeout;
```
If this is called multiple times, will there be a risk that it can always return a positive remaining time even if the elapsed time is longer than `timeout`?

rohan-varma replied:
Yes, in the case of `waitAllRanks`, we don't decrease the overall timeout after every call to `wait()`. The reasoning is explained in a comment in `monitoredBarrier`, but essentially if rank n times out, and we want to collect failure info for ranks > n, then the timeout when waiting for ranks > n would always be 0 if a previous rank timed out. Hence we always use the initial timeout for all calls to wait() when `waitAllRanks=true`.

SciPioneer commented on torch/lib/c10d/ProcessGroupGloo.hpp:
```diff
@@ -251,8 +251,13 @@ class ProcessGroupGloo : public ProcessGroup {
   void monitoredBarrier(
-      const BarrierOptions& opts = BarrierOptions()) override;
+      const BarrierOptions& opts = BarrierOptions(),
+      bool waitAllRanks = false) override;
```
I don't see an easy way to flip this in practice. What about automatically turning on this whenever we use a debug mode (e.g., `NCCL_DEBUG=INFO`)

rohan-varma replied:
This PR exposes this flag to the python interface, so users can just do `dist.monitored_barrier(wait_all_ranks=True)`. Although agreed, it is important to think about the default since it may be more useful to collect _all_ ranks if user is using this debug tool. Since this is mostly going to be used in conjunction with torch_distributed_debug, maybe wait_all_ranks=True by default makes sense, though wait_all_ranks=False provides a "fail fast" option.

rohan-varma replied:
My bad, I didn't realize it wasn't exposed to python in the version you reviewed. Added that now.

SciPioneer replied:
> This PR exposes this flag to the python interface, so users can just do `dist.monitored_barrier(wait_all_ranks=True)`. Although agreed, it is important to think about the default since it may be more useful to collect _all_ ranks if user is using this debug tool. Since this is mostly going to be used in conjunction with torch_distributed_debug, maybe wait_all_ranks=True by default makes sense, though wait_all_ranks=False provides a "fail fast" option.

We don't know what kind of error the user is going to investigate, aka., whether collecting all the ranks here will help. You can make a decision here based on which case is more common.

rohan-varma replied:
Good point, will land this change and spend some time thinking about the default after collecting some user feedback.

SciPioneer commented on torch/testing/_internal/distributed/distributed_test.py:
```diff
@@ -5211,16 +5207,38 @@ def test_monitored_barrier_allreduce_hang(self):
+        @skip_if_lt_x_gpu(int(os.environ["WORLD_SIZE"]))
+        def test_monitored_barrier_allreduce_hang_wait_all_ranks(self):
+            # tests expected behavior when nonzero rank hangs and we want to
+            # report first timed out rank.
```
Please update the comment.

SciPioneer commented on torch/lib/c10d/ProcessGroupGloo.cpp:
```diff
@@ -2756,36 +2763,73 @@ void ProcessGroupGloo::monitoredBarrier(const BarrierOptions& opts) {
+    }
+
+    if (failedRanks.size() > 0) {
+      auto ranksStr = c10::Join(", ", failedRanks);
```
Nit: use `cons std::string` instead of `auto`?

SciPioneer commented on torch/lib/c10d/ProcessGroupGloo.cpp:
```diff
@@ -2756,36 +2763,73 @@ void ProcessGroupGloo::monitoredBarrier(const BarrierOptions& opts) {
+      }
+    }
+
+    if (failedRanks.size() > 0) {
```
Nit: `! failedRanks.empty()` is usually preferred for readability.

rohan-varma committed:
Update on "[c10d] Monitored barrier: option to collect all failed ranks"


Follow up change to add a flag to provide an option for monitored barrier to collect all the failed ranks and then throw instead of just throwing on the first one. This is useful as now monitored barrier will be able to pick up on all hanging ranks instead of just one.

This is done by passing in a flag `wait_all_ranks=True`.

Differential Revision: [D27447787](https://our.internmc.facebook.com/intern/diff/D27447787/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[c10d] Monitored barrier: option to collect all failed ranks"


Follow up change to add a flag to provide an option for monitored barrier to collect all the failed ranks and then throw instead of just throwing on the first one. This is useful as now monitored barrier will be able to pick up on all hanging ranks instead of just one.

This is done by passing in a flag `wait_all_ranks=True`.

Differential Revision: [D27447787](https://our.internmc.facebook.com/intern/diff/D27447787/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[c10d] Monitored barrier: option to collect all failed ranks"


Follow up change to add a flag to provide an option for monitored barrier to collect all the failed ranks and then throw instead of just throwing on the first one. This is useful as now monitored barrier will be able to pick up on all hanging ranks instead of just one.

This is done by passing in a flag `wait_all_ranks=True`.

Differential Revision: [D27447787](https://our.internmc.facebook.com/intern/diff/D27447787/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[c10d] Monitored barrier: option to collect all failed ranks"


Follow up change to add a flag to provide an option for monitored barrier to collect all the failed ranks and then throw instead of just throwing on the first one. This is useful as now monitored barrier will be able to pick up on all hanging ranks instead of just one.

This is done by passing in a flag `wait_all_ranks=True`.

Differential Revision: [D27447787](https://our.internmc.facebook.com/intern/diff/D27447787/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[c10d] Monitored barrier: option to collect all failed ranks"


Follow up change to add a flag to provide an option for monitored barrier to collect all the failed ranks and then throw instead of just throwing on the first one. This is useful as now monitored barrier will be able to pick up on all hanging ranks instead of just one.

This is done by passing in a flag `wait_all_ranks=True`.

Differential Revision: [D27447787](https://our.internmc.facebook.com/intern/diff/D27447787/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[c10d] Monitored barrier: option to collect all failed ranks"


Follow up change to add a flag to provide an option for monitored barrier to collect all the failed ranks and then throw instead of just throwing on the first one. This is useful as now monitored barrier will be able to pick up on all hanging ranks instead of just one.

This is done by passing in a flag `wait_all_ranks=True`.

Differential Revision: [D27447787](https://our.internmc.facebook.com/intern/diff/D27447787/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[c10d] Monitored barrier: option to collect all failed ranks"


Follow up change to add a flag to provide an option for monitored barrier to collect all the failed ranks and then throw instead of just throwing on the first one. This is useful as now monitored barrier will be able to pick up on all hanging ranks instead of just one.

This is done by passing in a flag `wait_all_ranks=True`.

Differential Revision: [D27447787](https://our.internmc.facebook.com/intern/diff/D27447787/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[c10d] Monitored barrier: option to collect all failed ranks"


Follow up change to add a flag to provide an option for monitored barrier to collect all the failed ranks and then throw instead of just throwing on the first one. This is useful as now monitored barrier will be able to pick up on all hanging ranks instead of just one.

This is done by passing in a flag `wait_all_ranks=True`.

Differential Revision: [D27447787](https://our.internmc.facebook.com/intern/diff/D27447787/)

[ghstack-poisoned]
